src/internal/task/task.go

```diff
@@ -0,0 +1,69 @@
+// The input / output of a task should be fully specifiable by an Any proto.
+// The processing of a task should produce a single output Any or an error.
+// A task when run may have side effects, but is idempotent.
+// A task can be run 0 or more times.
+// A task can be run 0 times when a task with the same description (hash of the input Any) has been
+// processed already (only applicable when using a caching layer).
+// A task can be run multiple times if issues outside of the task processing occur.
+// A task must allow for two or more workers to process it concurrently and perform side effects correctly.
+// Tasks are managed by Doers and provided to client workers for processing through Sources.
+// NOTE: In general, tasks should be designed to take at least seconds and ideally minutes because
+// that is usually a good balance of being long enough to minimize the overhead of laying out the
+// work and not being too long to restart from scratch.
+
+// Namespace:
+// A task managed by a Service has a namespace.
+// The namespace is used for scoping the Doers and Sources of tasks such that clients can create workers
+// that only process tasks created by clients in the same namespace.
+
+// Scheduling:
+// A task managed by a Service has a group.
+// The group is used for scheduling purposes.
+// Scheduling is based on maximizing fairness for the groups, and the schedulable unit is a task.
+type Service interface {
+	// NewDoer creates a Doer with the provided namespace and group.
+	NewDoer(namespace, group string) Doer
+	// NewSource creates a Source with the provided namespace.
+	NewSource(namespace string) Source
+	// TaskCount returns how many tasks are in a namespace and how many are claimed.
+	TaskCount(ctx context.Context, namespace string) (tasks int64, claims int64, _ error)
+}
+
+// Doer is a doer of tasks.
+// Refer to the DoOne and DoBatch helper functions if a simpler interface is desired.
+type Doer interface {
+	// Do creates and returns the results of processing a stream of tasks provided
+	// by the input channel. The client should close the input channel when all tasks have
+	// been sent (it does not need to be closed if the context is canceled). For each
+	// task, the collect function will be called with the results.
+	Do(ctx context.Context, inputChan chan *types.Any, cb CollectFunc) error
+}
+
+// Source is a source of tasks.
+type Source interface {
+	// Iterate iterates through tasks until the provided context is canceled.
+	// For each task, the process function will be called and the results
+	// will be returned to the Doer that created the task.
+	Iterate(ctx context.Context, cb ProcessFunc) error
+}
+
+// CollectFunc is the type of a function that is used for collecting the output of a stream / batch of tasks.
+// Index is the index of a task with respect to the order in which the task was created in the stream / batch.
+type CollectFunc = func(index int64, output *types.Any, _ error) error
```

Comment:
This should probably be an `int` since it is going to be used as an index to a slice.

Reply:
I tend to use int64 for things that may need to support very high scale at some point. Although it probably won't be needed, the downside (casting to an int) doesn't really outweigh the potential upside (supporting higher scale) imo.